flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/InternalTimerServiceAsyncImpl.java

```diff
@@ -0,0 +1,146 @@
+            StreamTaskCancellationContext cancellationContext,
+            AsyncExecutionController<K> asyncExecutionController) {
+        super(
+                taskIOMetricGroup,
+                localKeyGroupRange,
+                keyContext,
+                processingTimeService,
+                processingTimeTimersQueue,
+                eventTimeTimersQueue,
+                cancellationContext);
+        this.asyncExecutionController = asyncExecutionController;
+        this.processingTimeCallback = this::onProcessingTime;
+    }
+
+    private void onProcessingTime(long time) throws Exception {
+        // null out the timer in case the Triggerable calls registerProcessingTimeTimer()
+        // inside the callback.
+        nextTimer = null;
+
+        InternalTimer<K, N> timer;
+
+        while ((timer = processingTimeTimersQueue.peek()) != null
+                && timer.getTimestamp() <= time
+                && !cancellationContext.isCancelled()) {
+            RecordContext<K> recordCtx =
+                    asyncExecutionController.buildContext(EMPTY_RECORD, timer.getKey());
+            recordCtx.retain();
+            asyncExecutionController.setCurrentContext(recordCtx);
+            keyContext.setCurrentKey(timer.getKey());
+            processingTimeTimersQueue.poll();
+            final InternalTimer<K, N> timerToTrigger = timer;
+            asyncExecutionController.syncPointRequestWithCallback(
+                    () -> triggerTarget.onProcessingTime(timerToTrigger));
+            taskIOMetricGroup.getNumFiredTimers().inc();
+            recordCtx.release();
+        }
+
+        if (timer != null && nextTimer == null) {
+            nextTimer =
+                    processingTimeService.registerTimer(
+                            timer.getTimestamp(), this::onProcessingTime);
+        }
+    }
+
+    /**
+     * Advance one watermark, this will fire some event timers.
+     *
+     * @param time the time in watermark.
+     */
+    @Override
+    public void advanceWatermark(long time) throws Exception {
+        currentWatermark = time;
+
+        InternalTimer<K, N> timer;
+
+        while ((timer = eventTimeTimersQueue.peek()) != null
+                && timer.getTimestamp() <= time
+                && !cancellationContext.isCancelled()) {
+            RecordContext<K> recordCtx =
+                    asyncExecutionController.buildContext(EMPTY_RECORD, timer.getKey());
+            recordCtx.retain();
+            asyncExecutionController.setCurrentContext(recordCtx);
+            keyContext.setCurrentKey(timer.getKey());
+            eventTimeTimersQueue.poll();
+            final InternalTimer<K, N> timerToTrigger = timer;
+            asyncExecutionController.syncPointRequestWithCallback(
+                    () -> triggerTarget.onEventTime(timerToTrigger));
+            taskIOMetricGroup.getNumFiredTimers().inc();
+            recordCtx.release();
+        }
+    }
+
+    protected void foreachTimer(
```

Comment:
Actually, providing batch operation in async execution is dangerous. This should be carefully evaluated. We could discuss this later.

Reply:
How about throw UnsupportedException for now and add actual implementations later after the evaluation is complete? It might be better not to support certain functions than to support risky functions.

Reply:
Agreed, let `foreachTimer` throw `UnsupportedOperationException` now.